networkx/algorithms/centrality/laplacian.py

```diff
@@ -0,0 +1,145 @@
+
+    alpha : real (default = 0.95)
+        Optional parameter for the Laplacian matrix calculation. (1 - alpha) is the teleportation probability used with pagerank.
+
+    Returns
+    -------
+    nodes : dictionary
+       Dictionary of nodes with Laplacian centrality as the value.
+
+    Examples
+    --------
+    >>> G = nx.Graph()
+    >>> edges = [(0, 1, 4), (0, 2, 2), (2, 1, 1), (1, 3, 2), (1, 4, 2), (4, 5, 1)]
+    >>> G.add_weighted_edges_from(edges)
+    >>> sorted((v, f"{c:0.2f}") for v, c in laplacian_centrality(G).items())
+    [(0, '0.70'), (1, '0.90'), (2, '0.28'), (3, '0.22'), (4, '0.26'), (5, '0.04')]
+
+    Notes
+    -----
+    The algorithm is implemented based on [1] with an extension to directed graphs
+    using the `nx.directed_laplacian_matrix` function.
+
+    Raises
+    ------
+    NetworkXPointlessConcept
+        If the graph `G` is the null graph.
+
+    References
+    ----------
+    .. [1] Qi, X., Fuller, E., Wu, Q., Wu, Y., and Zhang, C.-Q. (2012).
+    Laplacian centrality: A new centrality measure for weighted networks.
+    Information Sciences, 194:240-253.
+    https://math.wvu.edu/~cqzhang/Publication-files/my-paper/INS-2012-Laplacian-W.pdf
+    """
+    import numpy as np
+    import scipy as sp
+    import scipy.linalg  # call as sp.linalg
+    import scipy.sparse  # call as sp.sparse
+
+    def eigh_f(A):
+        return sp.linalg.eigh(A.toarray(), eigvals_only=True)
+
+    if len(G) == 0:
+        raise nx.NetworkXPointlessConcept(
+            "cannot compute centrality for the null graph"
+        )
+
+    if nodelist != None:
+        nodeset = set(G.nbunch_iter(nodelist))
+        if len(nodeset) != len(nodelist):
+            raise nx.NetworkXError(
+                "nodelist contains duplicate nodes or nodes not in G"
+            )
+        full_nodelist = nodelist + [n for n in G if n not in nodeset]
+    else:
+        nodeset = set(G.nbunch_iter(nodelist))
+        full_nodelist = nodeset
+
+    if G.is_directed():
+        lap_matrix = sp.sparse.csr_matrix(
+            nx.directed_laplacian_matrix(G, full_nodelist, weight, walk_type, alpha)
+        )
+        eigh = eigh_f(lap_matrix)
+
+    else:
+        lap_matrix = nx.laplacian_matrix(G, full_nodelist, weight)
+        eigh = eigh_f(lap_matrix)
+
+    if normalized:
+        sum_of_full = np.power(eigh, 2).sum()
+    else:
+        sum_of_full = 1
+
+    # calculate laplace centrality
+    laplace_centralities_dict = {}
+    for i, node in enumerate(nodeset):
+
+        new_diag = lap_matrix.diagonal() - abs(lap_matrix.getcol(i).toarray().flatten())
+
+        # remove row and col i from lap_matrix
+        all_but_i = list(np.arange(lap_matrix.shape[0]))
+        all_but_i.remove(i)
+        A_2 = lap_matrix[all_but_i, :]
+        A_2 = A_2[:, all_but_i]
+
+        A_2.setdiag(np.r_[new_diag[:i], new_diag[i + 1 :]])
```

Comment:
Similarly this can be simplified with advanced indexing.

```suggestion
        A_2.setdiag(new_diag[all_but_i])
```